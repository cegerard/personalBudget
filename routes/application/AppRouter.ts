import express from 'express';

import homeRoute from './home/routes';
import budgetsRoute from './budgets/routes';
import expensesRoute from './expenses/routes';
import balanceRoute from './balance/routes';
import ExpenseController from './expenses/controllers';
import BudgetController from './budgets/controllers';
import HomeController from './home/controllers';
import Authentication from '../../lib/security/authentication';
import { BalanceController } from './balance/controller';

class AppRouter {
  public router;

  constructor(
    homeController: HomeController,
    budgetController: BudgetController,
    expenseController: ExpenseController,
    balanceController: BalanceController,
  ) {
    this.router = express.Router();

    this.router.use('/', homeRoute.init(homeController));
    this.router.use(
      '/budgets',
      Authentication.ensureAuthenticated,
      budgetsRoute.init(budgetController)
    );
    this.router.use(
      '/expenses',
      Authentication.ensureAuthenticated,
      expensesRoute.init(expenseController)
    );
    this.router.use(
      '/balance',
      Authentication.ensureAuthenticated,
      balanceRoute.init(balanceController)
    );
  }
}

export default AppRouter;
