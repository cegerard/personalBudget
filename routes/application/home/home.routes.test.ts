import StatusCodes from 'http-status-codes';
import request, { Response } from 'supertest';

import application from '../../../app';
import { authenticate } from '../../../test/authHelper';
import UserRepositoryStub from '../../../test/stubs/UserRepositoryStub';

const userRepository = application.userRepository as UserRepositoryStub;
let app: any;

describe('Home', () => {
  beforeEach(() => {
    userRepository.resetStore();
    app = application.app;
  });

  describe('GET /', () => {
    describe('without authentication', () => {
      it('should response with 200', async () => {
        await request(app).get('/').expect(StatusCodes.OK);
      });

      it('should render home page', async () => {
        await request(app)
          .get('/')
          .then((res) => {
            expect(res.text).toMatchSnapshot();
          });
      });
    });

    describe('with authentication', () => {
      beforeEach(async () => {
        app = await authenticate(application.app);
      });

      it('should response with 200', async () => {
        await app.get('/').expect(StatusCodes.MOVED_TEMPORARILY);
      });

      it('should render budgets page', async () => {
        await app.get('/').then((res: any) => {
          expect(res.text).toMatchSnapshot();
        });
      });
    });
  });

  describe('POST /login', () => {
    describe('when the user does not exists', () => {
      it('should response with 302', async () => {
        await request(app)
          .post('/login')
          .set('Content-Type', 'application/json')
          .send({ email: 'd2v@lop.fr', password: '123' })
          .expect(StatusCodes.MOVED_TEMPORARILY);
      });

      it('should render login page', async () => {
        await request(app)
          .post('/login')
          .set('Content-Type', 'application/json')
          .send({ email: 'd2v@lop.fr', password: '123' })
          .then((res: Response) => {
            expect(res.text).toMatchSnapshot();
          });
      });
    });

    describe('when the user exists', () => {
      const userEmail = 'admin@d2velop.fr';
      const userPassword = 'pass';


      describe('when the credentials are not valid', () => {
        it('should response with 302', async () => {
          await request(app)
            .post('/login')
            .set('Content-Type', 'application/json')
            .send({ email: userEmail, password: 'not_valid' })
            .expect(StatusCodes.MOVED_TEMPORARILY);
        });

        it('should render login page', async () => {
          await request(app)
            .post('/login')
            .set('Content-Type', 'application/json')
            .send({ email: userEmail, password: 'not_valid' })
            .then((res: Response) => {
              expect(res.text).toMatchSnapshot();
            });
        });
      });

      describe('when the credentials are valid', () => {
        it('should response with 302', async () => {
          await request(app)
            .post('/login')
            .set('Content-Type', 'application/json')
            .send({ email: userEmail, password: userPassword })
            .expect(StatusCodes.MOVED_TEMPORARILY);
        });

        it('should render the budget list page', async () => {
          await request(app)
            .post('/login')
            .set('Content-Type', 'application/json')
            .send({ email: userEmail, password: userPassword })
            .then((res: Response) => {
              expect(res.text).toMatchSnapshot();
            });
        });
      });
    });
  });

  describe('GET /sign_up', () => {
    it('should response with 200', async () => {
      await request(app).get('/').expect(StatusCodes.OK);
    });

    it('should render sign up page', async () => {
      await request(app)
        .get('/sign_up')
        .then((res) => {
          expect(res.text).toMatchSnapshot();
        });
    });
  });

  describe('POST /sign_up', () => {
    let newUser: any;

    beforeEach(() => {
      newUser = {
        firstName: 'to',
        lastName: 'to',
        email: 'toto@gmail.com',
        password: '1234567',
        password2: '1234567',
      };
    });

    describe('when the sign up is successful', () => {
      it('should response with 302', async () => {
        await request(app)
          .post('/sign_up')
          .set('Content-Type', 'application/json')
          .send(newUser)
          .expect(StatusCodes.MOVED_TEMPORARILY);
      });

      it('should render budgets list page', async () => {
        await request(app)
          .post('/sign_up')
          .set('Content-Type', 'application/json')
          .send(newUser)
          .then((res: Response) => {
            expect(res.text).toMatchSnapshot();
          });
      });
    });

    describe('when passwords does not match', () => {
      beforeEach(() => {
        newUser.password2 = 'do not match';
      });

      it('should response with 400', async () => {
        await request(app)
          .post('/sign_up')
          .set('Content-Type', 'application/json')
          .send(newUser)
          .expect(StatusCodes.MOVED_TEMPORARILY);
      });

      it('should render sign up page', async () => {
        await request(app)
          .post('/sign_up')
          .set('Content-Type', 'application/json')
          .send(newUser)
          .then((res: Response) => {
            expect(res.text).toMatchSnapshot();
          });
      });
    });

    describe('when password is empty', () => {
      beforeEach(() => {
        newUser.password = '';
        newUser.password2 = '';
      });

      it('should response with 400', async () => {
        await request(app)
          .post('/sign_up')
          .set('Content-Type', 'application/json')
          .send(newUser)
          .expect(StatusCodes.MOVED_TEMPORARILY);
      });

      it('should render sign up page', async () => {
        await request(app)
          .post('/sign_up')
          .set('Content-Type', 'application/json')
          .send(newUser)
          .then((res: Response) => {
            expect(res.text).toMatchSnapshot();
          });
      });
    });

    describe('when password is undefined', () => {
      beforeEach(() => {
        newUser.password = undefined;
        newUser.password2 = undefined;
      });

      it('should response with 400', async () => {
        await request(app)
          .post('/sign_up')
          .set('Content-Type', 'application/json')
          .send(newUser)
          .expect(StatusCodes.MOVED_TEMPORARILY);
      });

      it('should render sign up page', async () => {
        await request(app)
          .post('/sign_up')
          .set('Content-Type', 'application/json')
          .send(newUser)
          .then((res: Response) => {
            expect(res.text).toMatchSnapshot();
          });
      });
    });

    describe('when password is null', () => {
      beforeEach(() => {
        newUser.password = null;
        newUser.password2 = null;
      });

      it('should response with 400', async () => {
        await request(app)
          .post('/sign_up')
          .set('Content-Type', 'application/json')
          .send(newUser)
          .expect(StatusCodes.MOVED_TEMPORARILY);
      });

      it('should render sign up page', async () => {
        await request(app)
          .post('/sign_up')
          .set('Content-Type', 'application/json')
          .send(newUser)
          .then((res: Response) => {
            expect(res.text).toMatchSnapshot();
          });
      });
    });

    describe('when user already exists', () => {
      beforeEach(async () => {
        await userRepository.create(newUser);
      });

      it('should response with 409', async () => {
        await request(app)
          .post('/sign_up')
          .set('Content-Type', 'application/json')
          .send(newUser)
          .expect(StatusCodes.MOVED_TEMPORARILY);
      });

      it('should render sign up page', async () => {
        await request(app)
          .post('/sign_up')
          .set('Content-Type', 'application/json')
          .send(newUser)
          .then((res: Response) => {
            expect(res.text).toMatchSnapshot();
          });
      });
    });
  });
});
