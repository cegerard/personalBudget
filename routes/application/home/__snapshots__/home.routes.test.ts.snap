// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Home GET / with authentication should render budgets page 1`] = `"Found. Redirecting to /budgets"`;

exports[`Home GET / without authentication should render home page 1`] = `
"<!DOCTYPE html><html><head><title>Personal Budget</title><link rel=\\"stylesheet\\" href=\\"/stylesheets/style.css\\"><link rel=\\"stylesheet\\" href=\\"https://stackpath.bootstrapcdn.com/bootstrap/4.5.0/css/bootstrap.min.css\\" integrity=\\"sha384-9aIt2nRpC12Uk9gS9baDl411NQApFmC26EwAOH8WgZl5MYYxFfc+NcPb1dKGj7Sk\\" crossorigin=\\"anonymous\\"><link rel=\\"stylesheet\\" href=\\"https://use.fontawesome.com/releases/v5.7.0/css/all.css\\" crossorigin=\\"anonymous\\"></head><body class=\\"container\\"><nav class=\\"navbar navbar-expand-lg navbar-dark bg-dark\\"><a class=\\"navbar-brand\\" href=\\"/\\">Personal Budget</a><button class=\\"navbar-toggler\\" type=\\"button\\" data-toggle=\\"collapse\\" data-target=\\"#navbarSupportedContent\\" aria-controls=\\"navbarSupportedContent\\" aria-expanded=\\"false\\" aria-label=\\"Toggle navigation\\"><span class=\\"navbar-toggler-icon\\"></span></button><div class=\\"collapse navbar-collapse\\" id=\\"navbarSupportedContent\\"><ul class=\\"navbar-nav mr-auto\\"><li class=\\"nav-item\\"><a class=\\"nav-link\\" href=\\"/budgets\\">Budgets</a></li><li class=\\"nav-item\\"><a class=\\"nav-link\\" href=\\"/expenses\\">Expenses</a></li></ul></div></nav><section class=\\"vh-100\\"><div class=\\"container py-5 h-100\\"><div class=\\"row d-flex justify-content-center align-items-center h-100\\"><div class=\\"col-12 col-md-8 col-lg-6 col-xl-5\\"><div class=\\"card bg-dark text-white\\" style=\\"border-radius: 1rem;\\"><div class=\\"card-body p-5 text-center\\"><form class=\\"mb-md-5 mt-md-4 pb-5\\" action=\\"/login\\" method=\\"post\\"><h2 class=\\"fw-bold mb-2 text-uppercase\\">Login</h2><p class=\\"text-white-50 mb-5\\">Please enter your login and password!</p><div class=\\"form-outline form-white mb-4\\"><input class=\\"form-control form-control-lg\\" type=\\"email\\" id=\\"email\\" name=\\"email\\"><label class=\\"form-label\\" for=\\"email\\">Email</label></div><div class=\\"form-outline form-white mb-4\\"><input class=\\"form-control form-control-lg\\" type=\\"password\\" id=\\"pass\\" name=\\"password\\"><label class=\\"form-label\\" for=\\"pass\\">Password</label></div><button class=\\"btn btn-outline-light btn-lg px-5\\" type=\\"submit\\">Login</button></form><div>
<p class=\\"mb-0\\">Don't have an account? <a href=\\"/sign_up\\" class=\\"text-white-50 fw-bold\\">Sign Up</a></p>
</div></div></div></div></div></div></section></body><script src=\\"https://code.jquery.com/jquery-3.5.1.slim.min.js\\"></script><script src=\\"https://cdn.jsdelivr.net/npm/popper.js@1.16.1/dist/umd/popper.min.js\\"></script><script src=\\"https://stackpath.bootstrapcdn.com/bootstrap/4.5.2/js/bootstrap.min.js\\"></script><script src=\\"/javascripts/script.js\\"></script></html>"
`;

exports[`Home GET /sign_up should render sign up page 1`] = `"<!DOCTYPE html><html><head><title>Personal Budget</title><link rel=\\"stylesheet\\" href=\\"/stylesheets/style.css\\"><link rel=\\"stylesheet\\" href=\\"https://stackpath.bootstrapcdn.com/bootstrap/4.5.0/css/bootstrap.min.css\\" integrity=\\"sha384-9aIt2nRpC12Uk9gS9baDl411NQApFmC26EwAOH8WgZl5MYYxFfc+NcPb1dKGj7Sk\\" crossorigin=\\"anonymous\\"><link rel=\\"stylesheet\\" href=\\"https://use.fontawesome.com/releases/v5.7.0/css/all.css\\" crossorigin=\\"anonymous\\"></head><body class=\\"container\\"><nav class=\\"navbar navbar-expand-lg navbar-dark bg-dark\\"><a class=\\"navbar-brand\\" href=\\"/\\">Personal Budget</a><button class=\\"navbar-toggler\\" type=\\"button\\" data-toggle=\\"collapse\\" data-target=\\"#navbarSupportedContent\\" aria-controls=\\"navbarSupportedContent\\" aria-expanded=\\"false\\" aria-label=\\"Toggle navigation\\"><span class=\\"navbar-toggler-icon\\"></span></button><div class=\\"collapse navbar-collapse\\" id=\\"navbarSupportedContent\\"><ul class=\\"navbar-nav mr-auto\\"><li class=\\"nav-item\\"><a class=\\"nav-link\\" href=\\"/budgets\\">Budgets</a></li><li class=\\"nav-item\\"><a class=\\"nav-link\\" href=\\"/expenses\\">Expenses</a></li></ul></div></nav><section class=\\"vh-100\\"><div class=\\"container py-5 h-100\\"><div class=\\"row d-flex justify-content-center align-items-center h-100\\"><div class=\\"col-12 col-md-8 col-lg-6 col-xl-5\\"><div class=\\"card bg-dark text-white\\" style=\\"border-radius: 1rem;\\"><div class=\\"card-body p-5 text-center\\"><form class=\\"mb-md-5 mt-md-4 pb-5\\" action=\\"/sign_up\\" method=\\"post\\"><h2 class=\\"fw-bold mb-2 text-uppercase\\">Sign Up</h2><p class=\\"text-white-50 mb-5\\">Create an account to manage your budgets</p><div class=\\"form-outline form-white mb-4\\"><input class=\\"form-control form-control-lg\\" type=\\"text\\" id=\\"firstname\\" name=\\"firstName\\"><label class=\\"form-label\\" for=\\"firstname\\">First name</label><div class=\\"form-outline form-white mb-4\\"></div><input class=\\"form-control form-control-lg\\" type=\\"text\\" id=\\"lastname\\" name=\\"lastName\\"><label class=\\"form-label\\" for=\\"lastname\\">Last name</label></div><div class=\\"form-outline form-white mb-4\\"><input class=\\"form-control form-control-lg\\" type=\\"email\\" id=\\"email\\" name=\\"email\\"><label class=\\"form-label\\" for=\\"email\\">Email</label></div><div class=\\"form-outline form-white mb-4\\"><input class=\\"form-control form-control-lg\\" type=\\"password\\" id=\\"pass\\" name=\\"password\\"><label class=\\"form-label\\" for=\\"pass\\">Password</label></div><div class=\\"form-outline form-white mb-4\\"><input class=\\"form-control form-control-lg\\" type=\\"password\\" id=\\"pass2\\" name=\\"password2\\"><label class=\\"form-label\\" for=\\"pass2\\">Retype password</label></div><p class=\\"mb-3\\">Have already an account?  <a href=\\"/\\" class=\\"text-white-50 fw-bold\\"><u>Login here</u></a></p><button class=\\"btn btn-outline-light btn-lg px-5\\" type=\\"submit\\">Create Account</button></form></div></div></div></div></div></section></body><script src=\\"https://code.jquery.com/jquery-3.5.1.slim.min.js\\"></script><script src=\\"https://cdn.jsdelivr.net/npm/popper.js@1.16.1/dist/umd/popper.min.js\\"></script><script src=\\"https://stackpath.bootstrapcdn.com/bootstrap/4.5.2/js/bootstrap.min.js\\"></script><script src=\\"/javascripts/script.js\\"></script></html>"`;

exports[`Home POST /login when the user does not exists should render login page 1`] = `"Found. Redirecting to /"`;

exports[`Home POST /login when the user exists when the credentials are not valid should render login page 1`] = `"Found. Redirecting to /"`;

exports[`Home POST /login when the user exists when the credentials are valid should render the budget list page 1`] = `"Found. Redirecting to /budgets"`;

exports[`Home POST /sign_up when password is empty should render sign up page 1`] = `"Bad Request. Redirecting to /sign_up"`;

exports[`Home POST /sign_up when password is null should render sign up page 1`] = `"Bad Request. Redirecting to /sign_up"`;

exports[`Home POST /sign_up when password is undefined should render sign up page 1`] = `"Bad Request. Redirecting to /sign_up"`;

exports[`Home POST /sign_up when passwords does not match should render sign up page 1`] = `"Bad Request. Redirecting to /sign_up"`;

exports[`Home POST /sign_up when the sign up is successful should render budgets list page 1`] = `"Found. Redirecting to /budgets"`;

exports[`Home POST /sign_up when user already exists should render sign up page 1`] = `"Conflict. Redirecting to /sign_up"`;
